### PR TITLE
[HOP] Autograd implementation for switch

### PR DESCRIPTION
**Pull request by bohnstingl**
This PR adds the Autograd functionality for switch. It is tightly coupled to the implementation for `cond` and generalizes it to multiple branches. It also consolidates a common functionality for the bw between the two into `utils.py`.
However, there is a special handling for the `int -> SymInt` and the `None` return case for switch. cond would in principle also support those cases and they could be implemented in a separate PR if needed.

This change was authored with the assistance of an AI coding assistant (Claude).

cc @ydwu4 @lakshayg @galv

### CODE REVIEW

**Comment by zou3519** [comment]
@claude review this please and put the review inside a GitHub collapsible section

**Comment by bohnstingl** [comment]
I've integrated the feedback from 1) and we now use the pytree functionality in the forward and the backward.
Also, I've added a comment for 2), i.e., we already enforce in the `_validate_inputs` that the pytree of the branches is identical, thus the gradient mask `grad_input_is_tensor` will also be correct.

Point 3 is something that I also stumbled across, but I don't really see an easy way out here. In principle we could specialize the materialization in the backward to only the relevant branch using `ctx._index`, however, only during eager mode as the index is a real int there. In compile-mode, this is a SymInt and we would need to either do a data-dependent branch to resolve it or materialize every branch as we have it. The only additional trick that we can include is to integrate a fast-track for when index is int, as we then only need to materialize one backward branch and can route the gradients through it directly. WDYT?

File: torch/_higher_order_ops/switch.py [hunk 1]
```diff
@@ -213,10 +214,90 @@ def switch_op_dense(index, branches, operands):
     return branches[clamped_idx](*operands)
 
 
+def _branch_tensor_outputs_only(branch):
```
Comment (Skylion007) [inline torch/_higher_order_ops/switch.py]: Add some typing so it doesn't type erase the branch function? You can use a TypeVar or a Callable ParamSpec or even a TupleTypeVar

File: torch/_higher_order_ops/utils.py [hunk 2]
```diff
@@ -1424,3 +1424,38 @@ def filter_with_masks(data: list[torch.Tensor | None], masks: list[bool]):
 def fill_none_with_masks(data: list[torch.Tensor | None], masks: list[bool]):
     data_iter = iter(data)
     return [next(data_iter) if kept else None for kept in masks]
+
+
+def materialize_bw_fn_filter_non_tensor_grads(
+    bw_fn: Callable,
+    args: tuple[Any, ...],
```
Comment (Skylion007) [inline torch/_higher_order_ops/utils.py]: Isn't this just the args from the Callable? Use a ParamSpec.

File: torch/_higher_order_ops/switch.py (hunk 1 above)
Comment (bohnstingl) [reply torch/_higher_order_ops/switch.py]: I added type annotations

File: torch/_higher_order_ops/utils.py (hunk 2 above)
Comment (bohnstingl) [reply torch/_higher_order_ops/utils.py]: I added type annotations

File: test/functorch/test_control_flow.py [hunk 3]
```diff
@@ -1940,6 +1940,612 @@ def branch2(inp_x):
...
+            operands = ({"t": [a, {"b": b}, (c,)]},)
+            result = switch(torch.tensor(i), branches, operands)
+            self.assertEqual(result, fn(operands[0]))
+
+            grad_out = torch.ones_like(result)
+            grads = torch.autograd.grad(result, (a, b, c), grad_out)
+            expected_grads = torch.autograd.grad(fn(operands[0]), (a, b, c), grad_out)
+            self.assertEqual(expected_grads, grads)
+
+    @skipIfTorchDynamo("Skip due to graph break when run with dynamo")
```
Comment (bohnstingl) [inline test/functorch/test_control_flow.py:2088]: Ideally, I'd like to first integrate the deduplication fix for switch, see https://github.com/pytorch/pytorch/pull/188374. With this landed, we can remove these skips

File: test/functorch/test_control_flow.py [hunk 4]
```diff
@@ -1940,6 +1940,612 @@ def branch2(inp_x):
...
+            # so the gradient is still well-defined.
+            result = switch(i, branches, (x,))
+            self.assertEqual(result, fn(x))
+
+            grad_out = torch.ones_like(result)
+            grads = torch.autograd.grad(result, (x,), grad_out)
+            expected_grads = torch.autograd.grad(fn(x), (x,), grad_out)
+            self.assertEqual(expected_grads, grads)
+
+    def test_switch_autograd_constant_index_specialization(self):
```
Comment (ydwu4) [inline test/functorch/test_control_flow.py]: In principle, we should only add tests that would fail without the current changes. i feel the constant index thing should be tested elsewhere?

File: test/functorch/test_control_flow.py [hunk 5]
```diff
@@ -1940,6 +1940,612 @@ def branch2(inp_x):
...
+            x = torch.randn(4, requires_grad=True)
+            result = switch(torch.tensor(i), branches, (x,))
+            self.assertEqual(result, fn(x))
+
+            grad_out = torch.ones_like(result)
+            grads = torch.autograd.grad(result, (x,), grad_out)
+            expected_grads = torch.autograd.grad(fn(x), (x,), grad_out)
+            self.assertEqual(expected_grads, grads)
+
+    def test_switch_autograd_grad_through_params(self):
```
Comment (ydwu4) [inline test/functorch/test_control_flow.py:2144]: how many these tests are actually unique testing different things? can we do a de-dup and keep the maningful ones?

File: test/functorch/test_control_flow.py [hunk 6]
```diff
@@ -1940,6 +1940,612 @@ def branch2(inp_x):
...
+            x = torch.randn(4, requires_grad=True, device="cuda")
+            result = switch(torch.tensor(i, device="cuda"), branches, (x,))
+            self.assertEqual(result, fn(x))
+
+            grad_out = torch.ones_like(result)
+            grads = torch.autograd.grad(result, (x,), grad_out)
+            expected_grads = torch.autograd.grad(fn(x), (x,), grad_out)
+            self.assertEqual(expected_grads, grads)
+
+    def test_switch_autograd_cond_equivalence(self):
```
Comment (ydwu4) [inline test/functorch/test_control_flow.py]: i don't think this is needed.

File: torch/_higher_order_ops/cond.py [hunk 7]
```diff
@@ -347,35 +345,17 @@ def backward(ctx, *flat_grads):
...
+        true_bw_gm, grads_tensor_masks = materialize_bw_fn_filter_non_tensor_grads(
+            ctx._true_bw_fn,
             args,
             ctx._fw_include_key_set,
             ctx._fw_exclude_key_set,
-            force_enable_grad=True,
         )
-        false_bw_gm = materialize_as_graph(
-            create_fn_remove_none(ctx._false_bw_fn),
+        false_bw_gm, _ = materialize_bw_fn_filter_non_tensor_grads(
```
Comment (ydwu4) [inline torch/_higher_order_ops/cond.py]: i don't want to add another materialize_bw_fn_filter_non_tensor_grads, can we move create_fn_remove_none out then do switch as cond?

File: test/functorch/test_control_flow.py (hunk 4 above)
Comment (bohnstingl) [reply test/functorch/test_control_flow.py]: I consolidated the testcases and only kept the relevant ones.

File: torch/_higher_order_ops/cond.py (hunk 7 above)
Comment (bohnstingl) [reply torch/_higher_order_ops/cond.py]: Okay. I reverted and used the exact same procedure as cond does. With the only difference that switch filters out the None values in the forward path as well, because switch allows Nones to be returned in from the branches.

File: test/functorch/test_control_flow.py [hunk 8]
```diff
@@ -1940,6 +1940,358 @@ def branch2(inp_x):
...
+
+        branches = (branch0, branch1, branch2)
+
+        @torch.compile(backend="eager", fullgraph=True, dynamic=True)
+        def f(idx_tensor, x):
+            return switch(idx_tensor, branches, (x,))
+
+        for i, fn in enumerate(branches):
+            x = torch.randn(4, requires_grad=True)
+            result = f(torch.tensor(i), x)
```
Comment (aorenste) [inline test/functorch/test_control_flow.py]: [cleanup] Named `symint_index`, but no SymInt index is actually produced here: `torch.tensor(i)` reaches `switch` as a `Tensor` index (schema `Tensor index`, cf. `test_switch_gen_schema_*`), and `@torch.compile(dynamic=True)` symbolizes tensor shapes, not the scalar value. So this runs the same Tensor-index path as the other new tests and leaves the SymInt-index autograd path uncovered. Drive the index from a genuine SymInt source (e.g. a data-dependent `.size(0)`/`.item()`, as `test_switch_gen_schema_symint_inputs` constructs one) so `switch(SymInt index, ...)` and its backward are exercised.

File: test/functorch/test_control_flow.py [hunk 9]
```diff
@@ -1940,6 +1940,358 @@ def branch2(inp_x):
...
+        branches = (branch0, branch1)
+
+        def f(idx, x):
+            result = switch(idx, branches, (x,))
+            grad_out = torch.ones_like(result)
+            return torch.autograd.grad(result, (x,), grad_out)
+
+        x = torch.randn(4, requires_grad=True)
+        gm = make_fx(f)(torch.tensor(1), x)
+        switch_calls = [
```
Comment (aorenste) [inline test/functorch/test_control_flow.py:1982]: [cleanup] Minor: this counts nodes via the substring `"switch" in str(n.target)` and asserts an exact count of `2` -- both couple the test to the current lowering (a forward node plus a distinct backward `switch` node), so a backward that inlines the branch selection would flip this red while gradients stay correct. It is also the only new autograd test that never asserts on gradient values (`make_fx` traces but the grads are not compared). Prefer matching on `torch.ops.higher_order.switch` identity and adding a grad-value check.

**Review by aorenste — APPROVED** [review]
(Reviewed by me, assisted by AI)

File: test/functorch/test_control_flow.py (hunk 8 above)
Comment (bohnstingl) [reply test/functorch/test_control_flow.py]: I adjusted the test such that it uses a genuine SymInt

File: test/functorch/test_control_flow.py (hunk 9 above)
Comment (bohnstingl) [reply test/functorch/test_control_flow.py:1982]: Fixed, I changed the string-based search for a proper node-based matching.
